File: backend/app/datasources/source_health.py
"""数据源引用健康状态（列表/详情提示用）。"""

from __future__ import annotations

import uuid
from typing import Literal

from sqlalchemy.orm import Session

from app.datasources.models import DataSource
from app.ingestion.models import SyncJob
from app.metadata.dataset.models import DatasetRecord
from app.query.config_store.models import QueryConfigRecord

SourceHealth = Literal["active", "missing", "none"]


def resolve_datasource_health(session: Session, ds_id: uuid.UUID | None) -> SourceHealth:
    if ds_id is None:
        return "none"
    row = session.get(DataSource, ds_id)
    if row is None or row.deleted_at is not None:
        return "missing"
    return "active"


def resolve_dataset_source_health(session: Session, row: DatasetRecord) -> SourceHealth:
    ds_id = row.table_source_datasource_id
    if ds_id is None and row.bound_config_id is not None:
        cfg = session.get(QueryConfigRecord, row.bound_config_id)
        if cfg is not None and isinstance(cfg.payload, dict):
            raw = cfg.payload.get("dataSourceId")
            if raw:
                try:
                    ds_id = uuid.UUID(str(raw))
                except ValueError:
                    pass
    return resolve_datasource_health(session, ds_id)


def resolve_sync_job_source_health(session: Session, job: SyncJob) -> SourceHealth:
    return resolve_datasource_health(session, job.source_data_source_id)

File: fe/src/lib/sourceHealth.ts
export type SourceHealth = "active" | "missing" | "none";

export function isSourceUnavailable(health: SourceHealth | undefined | null): boolean {
  return health === "missing";
}

export function sourceHealthBadgeLabel(health: SourceHealth | undefined | null): string {
  if (health === "missing") return "数据源不可用";
  return "";
}

export function sourceHealthAlertTitle(health: SourceHealth | undefined | null): string {
  if (health === "missing") return "数据源不可用或已删除";
  return "";
}

export function sourceHealthAlertDescription(
  entity: "dataset" | "sync_job",
): string {
  if (entity === "dataset") {
    return "绑定的数据连接已删除或不可见，查询与同步可能失败。请重新选择数据源或删除该 Dataset。";
  }
  return "任务引用的数据连接已删除或不可见，无法继续同步。请编辑任务并重新选择数据源。";
}
